File: languages/julia/src/PositionTape.jl
module PositionTape

using SHA

export Generate, GenerateMarkerComplete, GetMarkerCompleteLength, Locate, Validate,
    FindTruncationPoint, FindFirstMismatch, BuildWindowIndex, LocateByHash, HashFragment,
    Mismatch, ValidationResult

const DEFAULT_SEARCH_LENGTH = 100_003
const INDEX_CACHE = Dict{Int,Dict{String,Vector{Int}}}()

struct Mismatch
    position::Int
    expected::Union{Char,Nothing}
    received::Union{Char,Nothing}
end

struct ValidationResult
    is_valid::Bool
    expected_length::Int
    received_length::Int
    truncation_point::Union{Int,Nothing}
    first_mismatch::Union{Mismatch,Nothing}
end

function _assert_non_negative(requested_length::Int)
    requested_length < 0 && throw(ArgumentError("length must be non-negative"))
end

function Generate(requested_length::Int)::String
    _assert_non_negative(requested_length)
    output = IOBuffer()
    cursor = 1

    while position(output) < requested_length
        if cursor % 10 == 0
            marker = string(cursor ÷ 10)
            remaining = requested_length - position(output)
            print(output, marker[1:min(lastindex(marker), remaining)])
            cursor += length(marker)
        else
            print(output, cursor % 10)
            cursor += 1
        end
    end

    return String(take!(output))
end

function GetMarkerCompleteLength(requested_length::Int)::Int
    _assert_non_negative(requested_length)
    cursor = 1

    while cursor <= requested_length
        if cursor % 10 == 0
            marker_length = length(string(cursor ÷ 10))
            marker_end = cursor + marker_length - 1
            requested_length < marker_end && return marker_end
            cursor += marker_length
        else
            cursor += 1
        end
    end

    return requested_length
end

GenerateMarkerComplete(requested_length::Int)::String = Generate(GetMarkerCompleteLength(requested_length))

function FindFirstMismatch(expected::AbstractString, received::AbstractString)::Union{Mismatch,Nothing}
    shared_length = min(length(expected), length(received))
    for index in 1:shared_length
        if expected[index] != received[index]
            return Mismatch(index, expected[index], received[index])
        end
    end

    length(expected) == length(received) && return nothing
    position = shared_length + 1
    return Mismatch(
        position,
        position <= length(expected) ? expected[position] : nothing,
        position <= length(received) ? received[position] : nothing,
    )
end

function Validate(receivedText::AbstractString, expectedLength::Int)::ValidationResult
    expected = Generate(expectedLength)
    mismatch = FindFirstMismatch(expected, receivedText)
    truncation_point = nothing

    if mismatch !== nothing && length(receivedText) < expectedLength && startswith(expected, receivedText)
        truncation_point = length(receivedText) + 1
    end

    return ValidationResult(mismatch === nothing, expectedLength, length(receivedText), truncation_point, mismatch)
end

function FindTruncationPoint(receivedText::AbstractString)::Int
    mismatch = FindFirstMismatch(Generate(length(receivedText)), receivedText)
    return mismatch === nothing ? length(receivedText) + 1 : mismatch.position
end

function Locate(fragment::AbstractString)::Int
    isempty(fragment) && return 1
    found = findfirst(fragment, Generate(DEFAULT_SEARCH_LENGTH))
    return found === nothing ? -1 : first(found)
end

HashFragment(fragment::AbstractString)::String = bytes2hex(sha256(Vector{UInt8}(codeunits(fragment))))

function BuildWindowIndex(windowSize::Int)::Dict{String,Vector{Int}}
    windowSize <= 0 && throw(ArgumentError("windowSize must be positive"))
    windowSize > DEFAULT_SEARCH_LENGTH && throw(ArgumentError("windowSize cannot exceed the default search length"))

    tape = Generate(DEFAULT_SEARCH_LENGTH)
    index = Dict{String,Vector{Int}}()
    for offset in 1:(length(tape) - windowSize + 1)
        hash = HashFragment(tape[offset:offset + windowSize - 1])
        push!(get!(index, hash, Int[]), offset)
    end
    return index
end

function LocateByHash(fragmentHash::AbstractString, windowSize::Int)::Vector{Int}
    normalized_hash = lowercase(strip(fragmentHash))
    index = get!(INDEX_CACHE, windowSize) do
        BuildWindowIndex(windowSize)
    end
    return copy(get(index, normalized_hash, Int[]))
end

end

File: languages/ruby/src/position_tape.rb
require "digest"

module PositionTape
  DEFAULT_SEARCH_LENGTH = 100_003
  @index_cache = {}

  Mismatch = Struct.new(:position, :expected, :received, keyword_init: true)
  ValidationResult = Struct.new(
    :is_valid,
    :expected_length,
    :received_length,
    :truncation_point,
    :first_mismatch,
    keyword_init: true
  )

  module_function

  def Generate(length)
    raise ArgumentError, "length must be non-negative" if length.negative?

    output = +""
    cursor = 1
    while output.length < length
      if (cursor % 10).zero?
        marker = (cursor / 10).to_s
        output << marker[0, length - output.length]
        cursor += marker.length
      else
        output << (cursor % 10).to_s
        cursor += 1
      end
    end
    output
  end

  def GetMarkerCompleteLength(length)
    raise ArgumentError, "length must be non-negative" if length.negative?

    cursor = 1
    while cursor <= length
      if (cursor % 10).zero?
        marker_length = (cursor / 10).to_s.length
        marker_end = cursor + marker_length - 1
        return marker_end if length < marker_end
        cursor += marker_length
      else
        cursor += 1
      end
    end
    length
  end

  def GenerateMarkerComplete(length)
    Generate(GetMarkerCompleteLength(length))
  end

  def FindFirstMismatch(expected, received)
    shared_length = [expected.length, received.length].min
    (0...shared_length).each do |index|
      next if expected[index] == received[index]

      return Mismatch.new(position: index + 1, expected: expected[index], received: received[index])
    end

    return nil if expected.length == received.length

    position = shared_length + 1
    Mismatch.new(
      position: position,
      expected: position <= expected.length ? expected[position - 1] : nil,
      received: position <= received.length ? received[position - 1] : nil
    )
  end

  def Validate(received_text, expected_length)
    expected = Generate(expected_length)
    mismatch = FindFirstMismatch(expected, received_text)
    truncation_point = nil
    if mismatch && received_text.length < expected_length && expected.start_with?(received_text)
      truncation_point = received_text.length + 1
    end

    ValidationResult.new(
      is_valid: mismatch.nil?,
      expected_length: expected_length,
      received_length: received_text.length,
      truncation_point: truncation_point,
      first_mismatch: mismatch
    )
  end

  def FindTruncationPoint(received_text)
    mismatch = FindFirstMismatch(Generate(received_text.length), received_text)
    mismatch ? mismatch.position : received_text.length + 1
  end

  def Locate(fragment)
    return 1 if fragment.empty?

    index = Generate(DEFAULT_SEARCH_LENGTH).index(fragment)
    index ? index + 1 : -1
  end

  def HashFragment(fragment)
    Digest::SHA256.hexdigest(fragment)
  end

  def BuildWindowIndex(window_size)
    raise ArgumentError, "windowSize must be positive" unless window_size.positive?
    raise ArgumentError, "windowSize cannot exceed the default search length" if window_size > DEFAULT_SEARCH_LENGTH

    tape = Generate(DEFAULT_SEARCH_LENGTH)
    index = Hash.new { |hash, key| hash[key] = [] }
    (0..(tape.length - window_size)).each do |offset|
      index[HashFragment(tape[offset, window_size])] << offset + 1
    end
    index
  end

  def LocateByHash(fragment_hash, window_size)
    normalized_hash = fragment_hash.strip.downcase
    @index_cache[window_size] ||= BuildWindowIndex(window_size)
    @index_cache[window_size].fetch(normalized_hash, []).dup
  end

  alias generate Generate
  alias generate_marker_complete GenerateMarkerComplete
  alias get_marker_complete_length GetMarkerCompleteLength
  alias find_first_mismatch FindFirstMismatch
  alias validate Validate
  alias find_truncation_point FindTruncationPoint
  alias locate Locate
  alias hash_fragment HashFragment
  alias build_window_index BuildWindowIndex
  alias locate_by_hash LocateByHash
end
